File: app/services/ai/business_confirmation.py
"""Business data confirmation helpers (SSE payload + message contract)."""
from __future__ import annotations

import json
import re
from contextvars import ContextVar
from typing import Any

BUSINESS_CONFIRMATION_TOOL_NAME = "request_user_confirmation"
BUSINESS_CONFIRMATION_MESSAGE_PREFIX = "【业务确认】"
BUSINESS_CONFIRMATION_CANCEL_MARKER = f"{BUSINESS_CONFIRMATION_MESSAGE_PREFIX}用户已取消"
BUSINESS_CONFIRMATION_CONFIRM_MARKER = f"{BUSINESS_CONFIRMATION_MESSAGE_PREFIX}用户已确定"

_cancel_gate_armed: ContextVar[bool] = ContextVar(
    "business_confirmation_cancel_gate",
    default=False,
)


_DATE_ONLY_RE = re.compile(r"^\d{4}[-/.]\d{1,2}[-/.]\d{1,2}$")
_CN_DATE_RE = re.compile(r"^\d{4}年\d{1,2}月\d{1,2}日$")
_DATETIME_RE = re.compile(
    r"^\d{4}[-/.]\d{1,2}[-/.]\d{1,2}[ T]\d{1,2}:\d{2}"
    r"(?::\d{2})?(?:\.\d+)?(?:Z|[+-]\d{2}:?\d{2})?$"
)
_DECLARED_VALUE_TYPES = frozenset({"boolean", "number", "text", "date", "datetime"})


def infer_confirmation_value_type(field: dict[str, Any] | None) -> str:
    """只认声明的 value_type，或值本身已是日期/日期时间形态。"""
    payload = field if isinstance(field, dict) else {}
    declared = str(payload.get("value_type") or "string").strip().lower()
    if declared in _DECLARED_VALUE_TYPES:
        return declared
    raw = payload.get("value")
    text = "" if raw is None else str(raw).strip()
    if _DATETIME_RE.match(text):
        return "datetime"
    if _DATE_ONLY_RE.match(text) or _CN_DATE_RE.match(text):
        return "date"
    return "string"


def normalize_confirmation_field_types(fields: list[Any] | None) -> list[dict[str, Any]]:
    normalized: list[dict[str, Any]] = []
    for field in fields or []:
        if not isinstance(field, dict):
            continue
        next_field = dict(field)
        next_field["value_type"] = infer_confirmation_value_type(next_field)
        normalized.append(next_field)
    return normalized


def is_business_confirmation_cancel_message(text: str | None) -> bool:
    return BUSINESS_CONFIRMATION_CANCEL_MARKER in str(text or "")


def is_business_confirmation_confirm_message(text: str | None) -> bool:
    return BUSINESS_CONFIRMATION_CONFIRM_MARKER in str(text or "")


def is_business_confirmation_receipt_message(text: str | None) -> bool:
    """True for confirm/cancel receipts produced by the business confirmation card."""
    return is_business_confirmation_cancel_message(text) or is_business_confirmation_confirm_message(
        text
    )


def arm_cancel_confirmation_gate(user_message: str | None) -> bool:
    """Arm per-turn gate when the latest user message is a cancel confirmation."""
    armed = is_business_confirmation_cancel_message(user_message)
    _cancel_gate_armed.set(armed)
    return armed


def is_cancel_confirmation_gate_armed() -> bool:
    return bool(_cancel_gate_armed.get())


def cancel_gate_block_payload() -> dict[str, Any]:
    return {
        "status": "error",
        "error": "business_confirmation_cancelled",
        "message": (
            "用户刚取消业务确认：禁止再次调用 request_user_confirmation，不要重新弹确认卡。"
            "请只用文字确认已取消，并询问用户是否修改后重试或放弃。"
        ),
    }


def _as_dict(value: Any) -> dict[str, Any] | None:
    if isinstance(value, dict):
        return value
    if isinstance(value, str):
        text = value.strip()
        if not text:
            return None
        try:
            parsed = json.loads(text)
        except json.JSONDecodeError:
            return None
        return parsed if isinstance(parsed, dict) else None
    return None


def parse_confirmation_tool_output(tool_output: Any) -> dict[str, Any] | None:
    """Parse successful tool output into a confirmation payload dict."""
    payload = _as_dict(tool_output)
    if not payload:
        # Nested text/data_blocks shape from some runners
        if isinstance(tool_output, dict) and "text" in tool_output:
            payload = _as_dict(tool_output.get("text"))
    if not payload:
        return None
    if payload.get("status") != "awaiting_user":
        return None
    confirmation_id = str(payload.get("confirmation_id") or "").strip()
    ui = payload.get("ui")
    if not confirmation_id or not isinstance(ui, dict):
        return None
    fields = ui.get("fields")
    if not isinstance(fields, list) or not fields:
        return None
    return payload


def build_business_confirmation_sse(
    *,
    tool_name: str,
    tool_output: Any,
    tool_call_id: str = "",
) -> dict[str, Any] | None:
    """Build frontend SSE event for a business confirmation card."""
    if tool_name != BUSINESS_CONFIRMATION_TOOL_NAME:
        return None
    if is_cancel_confirmation_gate_armed():
        return None
    payload = parse_confirmation_tool_output(tool_output)
    if not payload:
        return None
    ui = payload["ui"]
    return {
        "type": "business_confirmation",
        "tool_call_id": tool_call_id,
        "confirmation_id": str(payload["confirmation_id"]),
        "title": str(ui.get("title") or "请确认以下信息"),
        "summary": str(ui.get("summary") or ""),
        "fields": normalize_confirmation_field_types(ui.get("fields") or []),
        "confirm_label": str(ui.get("confirm_label") or "确定"),
        "cancel_label": str(ui.get("cancel_label") or "取消"),
        "risk_note": str(ui.get("risk_note") or ""),
        "status": "pending",
    }


def format_field_snapshot_lines(fields: list[dict[str, Any]]) -> list[str]:
    lines: list[str] = []
    for field in fields:
        if not isinstance(field, dict):
            continue
        key = str(field.get("key") or "").strip()
        label = str(field.get("label") or key or "字段").strip()
        value = field.get("value")
        if value is None:
            value_text = ""
        else:
            value_text = str(value)
        if key:
            lines.append(f"- {label} ({key}): {value_text}")
        else:
            lines.append(f"- {label}: {value_text}")
    return lines


def build_user_confirmation_message(
    *,
    confirmed: bool,
    confirmation_id: str,
    fields: list[dict[str, Any]],
) -> str:
    """Build the plain user message sent after confirm/cancel click."""
    snapshot = "\n".join(format_field_snapshot_lines(fields))
    cid = confirmation_id.strip() or "unknown"
    if confirmed:
        body = (
            f"{BUSINESS_CONFIRMATION_MESSAGE_PREFIX}用户已确定\n"
            f"confirmation_id: {cid}\n"
            "请根据以下已确认字段继续执行（如需写入请调用相应工具）：\n"
            f"{snapshot}"
        )
    else:
        body = (
            f"{BUSINESS_CONFIRMATION_MESSAGE_PREFIX}用户已取消\n"
            f"confirmation_id: {cid}\n"
            "请立即终止本次录入/变更："
            "不要调用写入类工具；"
            "禁止再次调用 request_user_confirmation（不要重新弹确认卡）。"
            "只用文字确认已取消，并询问用户是否修改后重试或放弃。"
            "仅当用户随后明确提供新的/修改后的数据并要求继续时，才可再次请求确认。\n"
            f"当时字段快照：\n{snapshot}"
        )
    return body.strip()

File: frontend/src/utils/businessConfirmation.ts
/** Business data confirmation card helpers (not tool-execution HITL). */

export type BusinessConfirmationValueType =
  | "string"
  | "number"
  | "boolean"
  | "text"
  | "date"
  | "datetime";

export interface BusinessConfirmationField {
  key: string;
  label: string;
  value: unknown;
  editable?: boolean;
  value_type?: BusinessConfirmationValueType;
}

const DATE_ONLY_RE = /^(\d{4})[-/.](\d{1,2})[-/.](\d{1,2})$/;
const CN_DATE_RE = /^(\d{4})年(\d{1,2})月(\d{1,2})日$/;
const DATETIME_RE =
  /^(\d{4})[-/.](\d{1,2})[-/.](\d{1,2})[ T](\d{1,2}):(\d{2})(?::(\d{2}))?(?:\.\d+)?(?:Z|[+-]\d{2}:?\d{2})?$/;

function pad2(value: number): string {
  return String(value).padStart(2, "0");
}

function parseConfirmationDateParts(value: unknown): {
  y: number;
  m: number;
  d: number;
  hh?: number;
  mm?: number;
} | null {
  const text = String(value ?? "").trim();
  if (!text) return null;
  const datetime = text.match(DATETIME_RE);
  if (datetime) {
    return {
      y: Number(datetime[1]),
      m: Number(datetime[2]),
      d: Number(datetime[3]),
      hh: Number(datetime[4]),
      mm: Number(datetime[5]),
    };
  }
  const iso = text.match(DATE_ONLY_RE);
  if (iso) {
    return { y: Number(iso[1]), m: Number(iso[2]), d: Number(iso[3]) };
  }
  const cn = text.match(CN_DATE_RE);
  if (cn) {
    return { y: Number(cn[1]), m: Number(cn[2]), d: Number(cn[3]) };
  }
  return null;
}

export function confirmationDateInputValue(value: unknown): string {
  const parsed = parseConfirmationDateParts(value);
  if (!parsed) return "";
  return `${parsed.y}-${pad2(parsed.m)}-${pad2(parsed.d)}`;
}

export function confirmationDateTimeInputValue(value: unknown): string {
  const parsed = parseConfirmationDateParts(value);
  if (!parsed) return "";
  return `${parsed.y}-${pad2(parsed.m)}-${pad2(parsed.d)}T${pad2(parsed.hh ?? 0)}:${pad2(parsed.mm ?? 0)}`;
}

export function confirmationDateTimeDisplayValue(value: unknown): string {
  const input = confirmationDateTimeInputValue(value);
  return input ? input.replace("T", " ") : "";
}

export function confirmationDateFromValue(value: unknown): Date | null {
  const parsed = parseConfirmationDateParts(value);
  if (!parsed) return null;
  return new Date(parsed.y, parsed.m - 1, parsed.d, parsed.hh ?? 0, parsed.mm ?? 0);
}

export function inferConfirmationValueType(
  field: Pick<BusinessConfirmationField, "value" | "value_type">,
): BusinessConfirmationValueType {
  const declared = field.value_type;
  if (
    declared === "boolean" ||
    declared === "number" ||
    declared === "text" ||
    declared === "date" ||
    declared === "datetime"
  ) {
    return declared;
  }
  const text = field.value === null || field.value === undefined ? "" : String(field.value).trim();
  if (DATETIME_RE.test(text)) return "datetime";
  if (DATE_ONLY_RE.test(text) || CN_DATE_RE.test(text)) return "date";
  return "string";
}

export interface BusinessConfirmationState {
  confirmation_id: string;
  tool_call_id?: string;
  title: string;
  summary?: string;
  fields: BusinessConfirmationField[];
  confirm_label: string;
  cancel_label: string;
  risk_note?: string;
  status: "pending" | "submitted" | "stale";
  decision?: "confirmed" | "cancelled";
}

export const BUSINESS_CONFIRMATION_MESSAGE_PREFIX = "【业务确认】";
export const BUSINESS_CONFIRMATION_CANCEL_MARKER = `${BUSINESS_CONFIRMATION_MESSAGE_PREFIX}用户已取消`;

export function isBusinessConfirmationCancelMessage(content: string | undefined | null): boolean {
  return String(content || "").includes(BUSINESS_CONFIRMATION_CANCEL_MARKER);
}

export function shouldSuppressBusinessConfirmation(
  messages?: Array<{ role?: string; content?: string }>,
): boolean {
  if (!messages?.length) return false;
  for (let i = messages.length - 1; i >= 0; i -= 1) {
    const msg = messages[i];
    const role = String(msg?.role || "").toLowerCase();
    if (role !== "user" && role !== "human") continue;
    return isBusinessConfirmationCancelMessage(msg?.content);
  }
  return false;
}

function asFields(raw: unknown): BusinessConfirmationField[] {
  if (!Array.isArray(raw)) return [];
  return raw
    .filter((item): item is Record<string, unknown> => !!item && typeof item === "object")
    .map((item) => {
      const field: BusinessConfirmationField = {
        key: String(item.key || "").trim(),
        label: String(item.label || item.key || "字段").trim(),
        value: item.value ?? "",
        editable: item.editable !== false,
        value_type: (item.value_type as BusinessConfirmationValueType) || "string",
      };
      field.value_type = inferConfirmationValueType(field);
      return field;
    })
    .filter((item) => item.key || item.label);
}

export function parseBusinessConfirmationEvent(
  data: Record<string, unknown>,
): BusinessConfirmationState | null {
  if (String(data.type || "") !== "business_confirmation") return null;
  const confirmationId = String(data.confirmation_id || "").trim();
  const fields = asFields(data.fields);
  if (!confirmationId || fields.length === 0) return null;
  return {
    confirmation_id: confirmationId,
    tool_call_id: data.tool_call_id ? String(data.tool_call_id) : undefined,
    title: String(data.title || "请确认以下信息"),
    summary: String(data.summary || ""),
    fields,
    confirm_label: String(data.confirm_label || "确定"),
    cancel_label: String(data.cancel_label || "取消"),
    risk_note: String(data.risk_note || ""),
    status: "pending",
  };
}

export function formatBusinessConfirmationSnapshot(
  fields: BusinessConfirmationField[],
): string {
  return fields
    .map((field) => {
      const label = field.label || field.key || "字段";
      const value =
        field.value === null || field.value === undefined ? "" : String(field.value);
      if (field.key) return `- ${label} (${field.key}): ${value}`;
      return `- ${label}: ${value}`;
    })
    .join("\n");
}

export function buildBusinessConfirmationUserMessage(
  confirmed: boolean,
  confirmationId: string,
  fields: BusinessConfirmationField[],
): string {
  const snapshot = formatBusinessConfirmationSnapshot(fields);
  const cid = confirmationId.trim() || "unknown";
  if (confirmed) {
    return [
      `${BUSINESS_CONFIRMATION_MESSAGE_PREFIX}用户已确定`,
      `confirmation_id: ${cid}`,
      "请根据以下已确认字段继续执行（如需写入请调用相应工具）：",
      snapshot,
    ].join("\n");
  }
  return [
    `${BUSINESS_CONFIRMATION_MESSAGE_PREFIX}用户已取消`,
    `confirmation_id: ${cid}`,
    "请立即终止本次录入/变更：不要调用写入类工具；禁止再次调用 request_user_confirmation（不要重新弹确认卡）。只用文字确认已取消，并询问用户是否修改后重试或放弃。仅当用户随后明确提供新的/修改后的数据并要求继续时，才可再次请求确认。",
    "当时字段快照：",
    snapshot,
  ].join("\n");
}

export function markOtherBusinessConfirmationsStale<
  T extends { businessConfirmation?: BusinessConfirmationState },
>(messages: T[], activeConfirmationId: string): void {
  for (const msg of messages) {
    const card = msg.businessConfirmation;
    if (!card || card.status !== "pending") continue;
    if (card.confirmation_id === activeConfirmationId) continue;
    card.status = "stale";
  }
}
